src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -269,48 +294,99 @@ public TypeSymbolWithAnnotations SetIsAnnotated(CSharpCompilation compilation)
+            TypeSymbol type = TypeSymbol.MergeNullability(typeSymbol, variance, out bool hadNestedMismatch);
             Debug.Assert((object)type != null);
             hadNullabilityMismatch = hadTopLevelMismatch | hadNestedMismatch;
-            return Create(type, isNullable, CustomModifiers);
+            return Create(type, nullableAnnotation, CustomModifiers);
         }
 
         /// <summary>
         /// Merges nullability.
         /// <paramref name="hadNullabilityMismatch"/> is true if there was conflict.
         /// </summary>
-        private static bool? MergeIsNullable(bool? a, bool? b, VarianceKind variance, out bool hadNullabilityMismatch)
+        private static NullableAnnotation MergeNullableAnnotation(TypeSymbol type, NullableAnnotation a, NullableAnnotation b, VarianceKind variance, out bool hadNullabilityMismatch)
         {
             hadNullabilityMismatch = false;
             if (a == b)
             {
                 return a;
             }
+
+            if (a.IsAnyNullable() && b.IsAnyNullable())
+            {
+                return NullableAnnotation.Nullable;
+            }
+
             switch (variance)
             {
                 case VarianceKind.In:
-                    return (a == false || b == false) ? (bool?)false : null;
+
+                    if (a == NullableAnnotation.NotNullableBasedOnAnalysis || b == NullableAnnotation.NotNullableBasedOnAnalysis)
+                    {
+                        return NullableAnnotation.NotNullableBasedOnAnalysis;
+                    }
+
+                    if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
+                    {
+                        return NullableAnnotation.NotNullable;
+                    }
+
+                    Debug.Assert(a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown);
+                    return NullableAnnotation.Unknown;
+
                 case VarianceKind.Out:
-                    return (a == true || b == true) ? (bool?)true : null;
+                    if (a.IsAnyNullable())
+                    {
+                        Debug.Assert(!b.IsAnyNullable());
+                        return a;
+                    }
+
+                    if (b.IsAnyNullable())
+                    {
+                        return b;
+                    }
+
+                    if (a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown)
+                    {
+                        return NullableAnnotation.Unknown;
+                    }
+
+                    Debug.Assert((a == NullableAnnotation.NotNullable && b == NullableAnnotation.NotNullableBasedOnAnalysis) ||
+                        (b == NullableAnnotation.NotNullable && a == NullableAnnotation.NotNullableBasedOnAnalysis));
+                    return NullableAnnotation.NotNullable;
+
                 default:
```

Comment:
>default: [](start = 16, length = 8)

Similarly, readability would be improved here by sticking to the structure of the spec:
 - if either `n` or `m` are non-nullable, non-nullable. In this case, if the other is nullable, a warning should be produced.
 - if either `n` or `m` are nullable, nullable.
 - otherwise oblivious.

I understand this problem didn't originate with your PR though, but it became more visible with new added code.

Reply:
I'm giving you feedback that this is not simple to read and map to the spec (which defines the correct behavior). As it stands, I'm not sure that this case implements the correct behavior and it feels more trouble to do an exhaustive verification rather than improve the readability, especially when you account for future readers who will have the same questions looking at this.
I believe the downside of a couple more lines of codes is outweighted by having straightforward and maintainable rather than clever code. Is it much trouble to update?

---
In reply to: [231334073](https://github.com/dotnet/roslyn/pull/30913#discussion_r231334073) [](ancestors = 231334073)

Reply:
> Is it much trouble to update?

Any code change is not free. I do not believe it is worth spending time to rewriting it, especially that it make code bigger.

---
In reply to: [231339910](https://github.com/dotnet/roslyn/pull/30913#discussion_r231339910) [](ancestors = 231339910,231334073)